### ISSUE TITLE
Projected lights as shaders

### ISSUE THREAD

**Issue by angelusmetal** [angelusmetal/dungeon#123]
Right now, lights are being rendered as follows:
- A light texture (usually a radial gradient, but can be anything) gets scaled to the light's radius/diameter and rendered in a separate, viewport-sized buffer
- All entities that intersect with it which can project shadows are identified; for each of them, a black polygon gets rendered for each line that blocks light from the center towards infinity
- After a light has been cut around all light-blocking lines, it gets "merged" into another buffer that contains all lights, by blending additively (each light will at most add more light, never subtract)
- After all lights within the viewport are rendered, the light buffer gets blended with the scene buffer.

The above actually happens twice. Once it is rendered with shadows, and this affects the floor tiles. The second time, lights are rendered again but without "cutting" shadows (only the light textures). This shadow-less buffer gets applied to all the walls and entities. This is a crude approximation so that a character doesn't get darkened by shadows behind them.

Cutting to the point, we need to explore implementing the whole process with shaders:
- The fragment shader will return the value for each pixel
- One possibility is providing all the light data, as well as well as all the geometry data, to the shader, and have it calculate the light for each pixel
- This will render a light texture that will get blended with the scene
- We need to consider whether values can go beyond 1.0 to produce an over-exposure effect
- Can we do blending differently? E.g. having an unlit-scene buffer, a light-buffer and combining them with a shader instead of through a blend function
- Is there anything we can do to do a single pass rendering of light? For that we'd need to take entity silhouettes into account so that shadows don't occlude stuff in front of them

**Comment by angelusmetal** [comment]
See: https://stackoverflow.com/questions/24855256/how-to-output-fragment-values-greater-than-1-in-webgl

**Comment by angelusmetal** [comment]
I did a lot of work around this, setting up standalone main classes within com.dungeon.desktop.shader package for playing with shaders and experimenting

**MultiSampleShadowCastTest** 
This was a first attempt at doing shader-based lighting and shadows. First I wrote a simple shader to determine the luminosity of each pixel based on the distance of the pixel being rendered and the light source position, having some parameters into account (such as "hardness" to control how light decays, and "range" to control how far from the source will light reach).

Then I worked with the shadows. I attempted to determine shadows directly on the shader.
The geometry segments (each one is 2 pairs of x,y coordinates), are passed to fragment shaders (i.e. each pixel), which checks whether there any segment intersects the line from the current pixel being rendered to the light source. If there's no intersection, the luminosity value is returned, multiplied by the light color, and adding the "ambient", shadow color, into account in the blend (so the light color slowly falls into the shadow color; not necessarily black).

Then I repeated this within a loop with a slight offset zero or more times (depending on the desired samples) to achieve soft lights. I added the concept of light radius, which is the physical radius of the light source (e.g. size of the light bulb itself, not how far it reaches), so all additional samples are projected from within the source's radius.

This has shown to be really expensive, because all the intersections need to be recalculated on every pixel of every frame. Plus, adding samples to achieve good looking shadows linearly degrades overall performance to a crawl. But it helped me grasp a lot better about shader performance, and how it rapidly degrades with loops and conditionals.

**SimpleShadowCastTest**
Based on my previous finding, I decided to revisit a similar approach to the one I had in the game, but with much cleaner code and a couple of twists based on my newfound knowledge.

The original approach was to draw each light into a separate buffer and "cut" away all the occluded segments (which generated hard shadows with noticeable aliasing), before blending that buffer into the "all lights" buffer. I start with clearing the "current light" buffer with black. Then I draw the non-occluded light (using the light shader, but could also be a texture) and then I project from the light into each segment. This would normally generate 2 black triangles that need to be drawn on top of the current light.

The first "twist" is that I wrote an umbra/penumbra shader to cheaply and precisely draw soft shadows. I first attempted at doing this degrade by manipulating vertex colors, but then I realized that while one edge of the triangle was completely dark and the opposing one, completely lit; there was no way of telling which should be the vertex color for the shared one between both edges. The way I fixed this is making a shader that does the degrade itself and compensates the grading scale so it shows up correctly in a triangle.

But also, I needed to add back this concept of light radius so I could have offset light origins for the same light; e.g. the "left-most" and "right-most" point of the light-bulb rim. This small difference (and all the light points in between) is what makes lights soft. This allowed me to draw 4 triangles instead of 2. The 2 middle ones would be completely dark (umbra) while the 2 at the sides would transition from complete dark to complete transparent (penumbra). Each intersecting segment generates a similar set of 4 triangles and they all get rendered with the same penumbra shader.

**NormalMappingTest**
A different test I did was playing with normal mapping; I was curious to understand this for a while, so I started digging some material from the web, grabbed a texture to play with and wrote a shader (yes, I had to copy/past an example and fiddle with it until it finally began sinking into my skull). Once it clicked and I was able to make it work, I integrated it with `SimpleShadowCastTest`.

I also added some controller code to be able to add/remove lights, move/zoom the camera and create new geometry using the mouse. The final touch was doing some cleanup and packaging this into a coherent class for rendering lights: `LightsRenderer`. It would create and manage its own frame buffers and sprite batches, load the appropriate shaders and try its best to do optimal calls to the OpenGL wrapper; now that I better understand, I'm tossing away some of the abstractions that I built to help myself, since they end up making a lot of unnecessary calls and flushes.
